### ISSUE TITLE
ButchNormalization fails when nvidia GPU is used and the training size and the batch size is "well set"

### ISSUE THREAD

**Issue by enyecz** [tensorflow/tensorflow#46205]
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 20.04
- TensorFlow installed from (source or binary): binary installed by pip3
- TensorFlow version (use command below): 2.4.0
- Python version: 3.8.5
- CUDA/cuDNN version: 11.0/8.0
- GPU model and memory: Nvidia 1080Ti 11GB

**Describe the current behavior**
When the size of training set and the batch size is set so that the last batch contains only one element (i.e. len(training_set)%batch_size == 1 ), one of the weights of the batch normalization (the variance) is set to nan. The problem disappears when using CPU.

**Describe the expected behavior**
Have valid weights computed regardless of the size of the training set.

**Standalone code to reproduce the issue**
from tensorflow import keras                                                                                                                                                                                       
import numpy as np                                                                                                                                                                                                 
                                                                                                                                                                                                                   
inp = keras.layers.Input((1,1,1))                                                                                                                                                                                  
mid = keras.layers.Conv2D(1, (1,1))(inp)                                                                                                                                                                           
out = keras.layers.BatchNormalization()(mid)                                                                                                                                                                       
                                                                                                                                                                                                                   
mod = keras.models.Model(inputs=inp, outputs=out)                                                                                                                                                                  
mod.compile(optimizer='adam', loss='mse')                                                                                                                                                                          
                                                                                                                                                                                                                   
data = np.reshape(np.arange(9), (9,1,1,1))                                                                                                                                                                         
                                                                                                                                                                                                                   
mod.fit(data,data,batch_size=8)                                                                                                                                                                                    
print(mod.layers[2].get_weights()[3])             #This is the last layer's last weight; this is NAN with GPU

**Comment by ravikyram** [comment]
I have tried in colab with TF GPU version 2.4 and i am not seeing any issue. Please, find the gist [here](https://colab.research.google.com/gist/ravikyram/949502c0c9083531adec9df661b04f01/untitled599.ipynb).Colab is using Cuda 10.1.Thanks!

**Comment by enyecz** [comment]
Strange, I just rerun the same box and got this:

2/2 [==============================] - 0s 5ms/step - loss: 19.9350
[nan]

**Comment by jvishnuvardhan** [comment]
@enyecz I ran @ravikyram gist and saw a number (instead of nan as you mentioned). Please check the [gist here](https://colab.research.google.com/gist/jvishnuvardhan/062e6c781ea27e283eab3636e423fc31/untitled599.ipynb)

Did you ran the code with (i) colab, (ii) local with CUDA 11.0, and (iii) both colab & local? Thanks!

**Comment by enyecz** [comment]
It seems to me that collab don't like pip. If you run pip first to install tensorflow-gpu, you don't get a GPU later, and running by CPU, there is no bug. In that case, you need to wait for the timeout, and your GPU-s come back with TF 2.4.0 (that's why I saw nan at the very same notebook).
Please take a look at the video I did, there I show you the whole case (I pushed yes on restarting the runtime, the transcoding went wrong somehow). There I use tf.test.is_gpu_available() to show that the GPU is lost.

https://user-images.githubusercontent.com/14057706/103832769-7477a980-507f-11eb-9da7-f7d143dc5109.mp4

**Comment by jvishnuvardhan** [comment]
@enyecz Agree with you. I just tried with `TF2.3-gpu`, which results in `nan`.  Please check the [gist here](https://colab.research.google.com/gist/jvishnuvardhan/0a952a327962dd10f7d07ec03c1e0704/untitled599.ipynb)

There is some issue in `TF2.4-gpu` when pip installed in a colab. We will look into the issue. Thanks!

**Comment by sushreebarsa** [comment]
Was able to reproduce the issue with TF v2.5-gpu,please find the gist [here](https://colab.research.google.com/gist/sushreebarsa/4a89ff3d35fac190ffd204cbe6c1f015/untitled599.ipynb#scrollTo=mdDdU-pI-J-P) ..Thanks!

**Comment by mohantym** [comment]
It is still replicating in TF 2.7. Attaching [Gist](https://colab.research.google.com/gist/mohantym/acebe3e0712983a629e878d4e5ef1b9f/untitled599.ipynb#scrollTo=Ld1i7Dou-umj) for reference. Thanks!

**Comment by enyecz** [comment]
It says:

2/2 [==============================] - 13s 26ms/step - loss: 18.6861
[nan]

So yes. This is still a bug.

**Comment by enyecz** [comment]
BUMP: The bug is still there with TF 2.8.2. I faced with it in a new project (and lost half a hour before I recognized the problem and had a facepalm). How can this not be a problem for others?

**Comment by Venkat6871** [comment]
Hi **@enyecz** ,
I believe this issue has been resolved in recent versions. I tried running your code on Colab using TensorFlow v2.17.0, and I did not encounter any issues. Please find the [gist](https://colab.research.google.com/gist/Venkat6871/a8c06564511c9af3788f5fcf5804f695/46205_tf-2-17-0-v.ipynb) here for your reference.
Thank you!
